### ISSUE TITLE
Improve the way the zt_channel_get and zt_channel_set are called

### ISSUE THREAD

**Issue by rodrigopex** [zeta-middleware/zeta#29]
### Current implementation
```c
u8_t some_channel = 0;
zt_channel_get(ZETA_SOME_CHANNEL, &some_channel, sizeof(some_channel));
printk("Some channel is %u\n", some_channel);
some_channel = 100;
zt_channel_set(ZETA_SOME_CHANNEL, &some_channel, sizeof(some_channel));
```
### ~~Proposal (adition)~~
```c
u8_t some_channel = 0;
ZT_CGET(ZETA_SOME_CHANNEL, some_channel);
printk("Some channel is %u\n", some_channel);
some_channel = 100;
ZT_CSET(ZETA_SOME_CHANNEL, some_channel);
```
~~Note the former implementation will not be removed, it will still work. The proposal is only to add an abstraction.~~

### Proposal

The proposed change seems to be a good way to handle channels. It is a crucial change to the project. The solution will help developers to write code easily. The only concern here is the usage of the GCC  extensions _Compound Literals_ and _Flexible Arrays_. For now, thinking of zephyr it will not be a big deal, but in the future when we'll port this to other RTOSes the problem will emerge. Let's keep it in mind. 

```c
#include <stdio.h>
#include <string.h>

typedef struct {
    size_t size;
    s8_t value;
} zt_data_s8_t;

typedef struct {
    size_t size;
    u8_t value;
} zt_data_u8_t;

typedef struct {
    size_t size;
    s16_t value;
} zt_data_s16_t;

typedef struct {
    size_t size;
    u16_t value;
} zt_data_u16_t;

typedef struct {
    size_t size;
    s32_t value;
} zt_data_s32_t;

typedef struct {
    size_t size;
    u32_t value;
} zt_data_u32_t;

typedef struct {
    size_t size;
    s64_t value;
} zt_data_s64_t;

typedef struct {
    size_t size;
    u64_t value;
} zt_data_u64_t;

typedef struct {
    size_t size;
    u8_t value[];
} zt_data_bytes_t;

union data {
    zt_data_s8_t s8;
    zt_data_u8_t u8;
    zt_data_s16_t s16;
    zt_data_u16_t u16;
    zt_data_s32_t s32;
    zt_data_u32_t u32;
    zt_data_s64_t s64;
    zt_data_u64_t u64;
    zt_data_bytes_t bytes;
};

typedef union data zt_data_t;

void zt_channel_set(u8_t channel, zt_data_t* data)
{
    printf("Data size %lu bytes\n", data->bytes.size);
    printf("Set channel %u bytes\n", channel);
    for (size_t i = 0; i < data->bytes.size; i++) {
        printf("arr[%lu] = %x\n", i, data->bytes.value[i]);
    }
}

void zt_channel_get(u8_t channel, zt_data_t* data)
{
    printf("Data size %lu bytes\n", data->bytes.size);
    u8_t db[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    memcpy(data->bytes.value, db, data->bytes.size);
    printf("Get channel %u bytes\n", channel);
}

#define ZT_DATA_S8(data)          \
    (zt_data_t*) (zt_data_s8_t[]) \
    {                             \
        {                         \
            1, data               \
        }                         \
    }

#define ZT_DATA_U8(data)          \
    (zt_data_t*) (zt_data_u8_t[]) \
    {                             \
        {                         \
            1, data               \
        }                         \
    }

#define ZT_DATA_S16(data)          \
    (zt_data_t*) (zt_data_s16_t[]) \
    {                              \
        {                          \
            2, data                \
        }                          \
    }

#define ZT_DATA_U16(data)          \
    (zt_data_t*) (zt_data_u16_t[]) \
    {                              \
        {                          \
            2, data                \
        }                          \
    }

#define ZT_DATA_S32(data)          \
    (zt_data_t*) (zt_data_s32_t[]) \
    {                              \
        {                          \
            4, data                \
        }                          \
    }

#define ZT_DATA_U32(data)          \
    (zt_data_t*) (zt_data_u32_t[]) \
    {                              \
        {                          \
            4, data                \
        }                          \
    }

#define ZT_DATA_S64(data)          \
    (zt_data_t*) (zt_data_s64_t[]) \
    {                              \
        {                          \
            8, data                \
        }                          \
    }

#define ZT_DATA_U64(data)          \
    (zt_data_t*) (zt_data_u64_t[]) \
    {                              \
        {                          \
            8, data                \
        }                          \
    }

#define ZT_DATA_BYTES(_size, data, ...) \
    (zt_data_t*) (struct {              \
        size_t size;                    \
        u8_t value[_size];              \
    }[])                                \
    {                                   \
        {                               \
            _size,                      \
            {                           \
                data, ##__VA_ARGS__     \
            }                           \
        }                               \
    }

int main()
{
    zt_data_t* d1 = ZT_DATA_U64(0);
    zt_channel_get(11, d1);
    zt_channel_set(1, d1);
    printf("d1.u32.size = %lu\n", d1->u32.size);
    printf("d1.u8.value = %02x\n", d1->u8.value);
    printf("d1.u16.value = %04x\n", d1->u16.value);
    printf("d1.u32.value = %08x\n", d1->u32.value);
    printf("d1.u64.value = %016lx\n", d1->u64.value);

    zt_data_t* d2 = ZT_DATA_S64(-1);
    zt_channel_set(1, d2);
    printf("d2.s8.value = %d\n", d2->s8.value);
    printf("d2.u8.value = %u\n", d2->u8.value);
    printf("d2.s16.value = %d\n", d2->s16.value);
    printf("d2.u16.value = %u\n", d2->u16.value);
    printf("d2.s32.size = %ld\n", d2->s32.size);
    printf("d2.u32.size = %lu\n", d2->u32.size);
    printf("d2.s32.value = %d\n", d2->s32.value);
    printf("d2.u32.value = %u\n", d2->u32.value);
    printf("d2.s64.value = %ld\n", d2->s64.value);
    printf("d2.u64.value = %lu\n", d2->u64.value);

    zt_data_t* arr = ZT_DATA_BYTES(32, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
    zt_channel_set(2, arr);
    printf("arr.u32.size = %lu\n", arr->u32.size);
    printf("arr.u32.value = %x\n", arr->u32.value);
    printf("arr.u8.value = %x\n", arr->u8.value);

    zt_channel_set(2, ZT_DATA_S8(0x17));
    zt_channel_set(2, ZT_DATA_U8(0x17));
    zt_channel_set(3, ZT_DATA_S16(0x1203));
    zt_channel_set(3, ZT_DATA_U16(0x1203));
    zt_channel_set(4, ZT_DATA_S32(0x29385212));
    zt_channel_set(4, ZT_DATA_U32(0x29385212));
    zt_channel_set(5, ZT_DATA_S64(0x11231231545457));
    zt_channel_set(5, ZT_DATA_U64(0x11231231545457));
    return 0;
}

```

#### References
[[1](https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Compound-Literals.html#Compound-Literals)] - GCC compound literals
[[2](https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Zero-Length.html)] - Zero length Array (Flexible array)

**Comment by rodrigopex** [comment]
@lucaspeixotot what do you think?

**Comment by rodrigopex** [comment]
Resolved by #39
